File: src/bumperbot_py_examples/bumperbot_py_examples/simple_parametric.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rcl_interfaces.msg import SetParametersResult
from rclpy.parameter import Parameter
# create a parametric node class
class SimpleParametric(Node): # inherit from Node
    
    def __init__(self): # constructor
        
        # call constructor of the base class
        super().__init__('simple_parametric') 
        
        # declare and provide defaults for parameters 
        self.declare_parameter('simple_int_param', 0)
        self.declare_parameter('simple_string_param', 'hello')

        # add callback for parameter changes
        self.add_on_set_parameters_callback(self.parameterChangeCallback)

    # define callback as a member of the class
    def parameterChangeCallback(self, params): 
        result=SetParametersResult()
        for param in params:
            if param.name == 'simple_int_param' and param.type_ == Parameter.Type.INTEGER:
                self.get_logger().info('Parameter simple_int_param changed to : %d' 
                               % param.value)
                result.successful = True
            
            if param.name == 'simple_string_param' and param.type_ == Parameter.Type.STRING:
                self.get_logger().info('Parameter simple_string_param changed to : %s' 
                               % param.value)
                result.successful = True

        return result


def main():

    # 1) Initialization: calling `init` for a particular Context. 
    # Must be done before any ROS nodes can be created.
    rclpy.init()
    
    # 2) Create one or more ROS nodes: by calling `create_node` or instantiating
    # a Node. Nodes may implement common ROS entities such as publishers, 
    # subscriptions, services and actions.
    simple_parametric = SimpleParametric() # create a SimplePublisher object

    # 3) Process node callbacks: by calling `spin`, `spin_once`, or 
    # `spin_until_future_complete`.
    # After a node is created, one can process work items (e.g. subscription 
    # callbacks) that are waiting to be executed
    rclpy.spin(simple_parametric) # process callbacks

    # 4) Shutdown
    # When finished with a previously initialized .Context (i.e. done using
    # all ROS nodes associated with the context), the shutdown function should 
    # be called. This will invalidate all entities derived from the context.
    simple_parametric.destroy_node() # destroy node
    rclpy.shutdown() # shutdown ROS2

if __name__ == '__main__':
    main()

File: src/bumperbot_cpp_examples/src/simple_parametric.cpp
#include <rclcpp/rclcpp.hpp>

#include <string>
#include <vector>
#include <memory>

#include <rcl_interfaces/msg/set_parameters_result.hpp>

// Define the SimpleParametric class as a subclass of rclcpp::Node
class SimpleParametric : public rclcpp::Node
{
public:
    // constructor, initialize Node with node name "simple_parametric"
    SimpleParametric() : Node("simple_parametric")
    {
        // declare a parameter named "simple_int_param" of type int with default value 7
        declare_parameter<int>("simple_int_param", 7);

        // declare a parameter named "simple_string_param" of type string with default value "default"
        declare_parameter<std::string>("simple_string_param", "default");

        // add a callback function to be called when parameters are set
        param_callback_handle_ = add_on_set_parameters_callback(std::bind(&SimpleParametric::paramChangeCallback, this, std::placeholders::_1));

        // log a message of which parameters the node has started with 
        RCLCPP_INFO_STREAM(get_logger(), "Node started with simple_int_param = " 
        << get_parameter("simple_int_param").as_int() 
        << " and simple_string_param = " 
        << get_parameter("simple_string_param").as_string());
    }

private:
    // shared pointer to store the output of add_on_set_parameters_callback
    OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;

    // callback function to be called when parameters are set
    rcl_interfaces::msg::SetParametersResult paramChangeCallback(
        const std::vector<rclcpp::Parameter> &parameters)
    {
        rcl_interfaces::msg::SetParametersResult result;

        // loop through the parameters
        for (const auto &param : parameters)
        {
            if ( 
                param.get_name() == "simple_int_param" && 
                param.get_type() == rclcpp::ParameterType::PARAMETER_INTEGER)
            {
                RCLCPP_INFO_STREAM(get_logger(), "New value for simple_int_param: " << param.as_int());
                result.successful = true;
                result.reason = "simple_int_param updated";
            }

            else if (
                param.get_name() == "simple_string_param" &&
                param.get_type() == rclcpp::ParameterType::PARAMETER_STRING)
            {
                RCLCPP_INFO_STREAM(get_logger(), "New value for simple_string_param: " << param.as_string());
                result.successful = true;
                result.reason = "simple_string_param updated";
            }
            else
            {
                result.successful = false;
                result.reason = "Invalid parameter";
            }
        }

        return result;
    }
};

int main(int argc, char *argv[])
{
    // initialize ROS2 interface
    rclcpp::init(argc, argv);

    // create a shared pointer to the SimpleParametric class
    auto node = std::make_shared<SimpleParametric>();
    
    // run the node
    rclcpp::spin(node);

    // shutdown ROS2
    rclcpp::shutdown();

    return 0;
}